app/assets/javascripts/discourse/app/form-kit/lib/fk-form-data.js

```diff
@@ -0,0 +1,198 @@
+   */
+  get isInvalid() {
+    return !this.isValid;
+  }
+
+  /**
+   * Checks if the changeset is pristine.
+   * @return {boolean} True if no patches have been applied.
+   */
+  get isPristine() {
+    return this.patches.length + this.inversePatches.length === 0;
+  }
+
+  /**
+   * Checks if the changeset is dirty.
+   * @return {boolean} True if patches have been applied.
+   */
+  get isDirty() {
+    return !this.isPristine;
+  }
+
+  /**
+   * Executes the patches to update the data.
+   */
+  execute() {
+    this.data = applyPatches(this.data, this.patches);
+  }
+
+  /**
+   * Reverts the patches to update the data.
+   */
+  unexecute() {
+    this.data = applyPatches(this.data, this.inversePatches);
+  }
+
+  /**
+   * Saves the changes by executing the patches and resetting them.
+   */
+  save() {
+    this.execute();
+    this.resetPatches();
+  }
+
+  /**
+   * Rolls back all changes by applying the inverse patches.
+   * @return {Promise<void>} A promise that resolves after the rollback is complete.
+   */
+  async rollback() {
+    while (this.inversePatches.length > 0) {
+      this.draftData = applyPatches(this.draftData, [
+        this.inversePatches.pop(),
+      ]);
+    }
+
+    this.resetPatches();
+
+    await new Promise((resolve) => next(resolve));
+  }
+
+  /**
+   * Adds an error to a specific property.
+   * @param {string} name - The property name.
+   * @param {Object} error - The error to add.
+   * @param {string} error.title - The title of the error.
+   * @param {string} error.message - The message of the error.
+   */
+  addError(name, error) {
+    if (this.errors.hasOwnProperty(name)) {
+      this.errors[name].messages.push(error.message);
+      this.errors = { ...this.errors };
+    } else {
+      this.errors = {
+        ...this.errors,
+        [name]: {
+          title: error.title,
+          messages: [error.message],
+        },
+      };
+    }
+  }
+
+  /**
+   * Removes an error from a specific property.
+   * @param {string} name - The property name.
+   */
+  removeError(name) {
+    delete this.errors[name];
+    this.errors = { ...this.errors };
+  }
+
+  /**
+   * Removes all errors from the changeset.
+   */
+  removeErrors() {
+    this.errors = {};
+  }
+
+  /**
+   * Gets the value of a specific property from the draft data.
+   * @param {string} name - The property name.
+   * @return {any} The value of the property.
+   */
+  get(name) {
+    const parts = name.split(".");
+    let target = this.draftData[parts.shift()];
+    while (parts.length) {
+      target = target[parts.shift()];
+    }
+    return target;
+  }
+
+  /**
+   * Sets the value of a specific property in the draft data and tracks the changes.
+   * @param {string} name - The property name.
+   * @param {any} value - The value to set.
+   */
+  set(name, value) {
+    this.draftData = produce(
+      this.draftData,
+      (d) => {
+        set(d, name, value);
```

Comment:
Would be good to replace this Ember `set`, like we did for `get`

Reply:
ah yes good point

Reply:
This is now done